### PR TITLE
feat: Allow HTML responses from tool results to include a custom message

### PR DESCRIPTION
**Pull request by Haervwe**
# Pull Request Checklist

- [x] **Target branch:** Verify that the pull request targets the `dev` branch. **PRs targeting `main` will be immediately closed.**
- [x] **Description:** Provide a concise description of the changes made in this pull request down below.
- [x] **Changelog:** Ensure a changelog entry following the format of [Keep a Changelog](https://keepachangelog.com/) is added at the bottom of the PR description.
- [x] **Documentation:** Add docs in [Open WebUI Docs Repository](https://github.com/open-webui/docs). Document user-facing behavior, environment variables, public APIs/interfaces, or deployment steps.
- [x] **Dependencies:** Are there any new or upgraded dependencies? If so, explain why, update the changelog/docs, and include any compatibility notes. Actually run the code/function that uses updated library to ensure it doesn't crash.
- [x] **Testing:** Perform manual tests to **verify the implemented fix/feature works as intended AND does not break any other functionality**. Include reproducible steps to demonstrate the issue before the fix. Test edge cases (URL encoding, HTML entities, types). Take this as an opportunity to **make screenshots of the feature/fix and include them in the PR description**.
- [x] **Agentic AI Code:** Confirm this Pull Request is **not written by any AI Agent** or has at least **gone through additional human review AND manual testing**. If any AI Agent is the co-author of this PR, it may lead to immediate closure of the PR.
- [x] **Code review:** Have you performed a self-review of your code, addressing any coding standard issues and ensuring adherence to the project's coding standards?
- [x] **Design & Architecture:** Prefer smart defaults over adding new settings; use local state for ephemeral UI logic. Open a Discussion for major architectural or UX changes.
- [x] **Git Hygiene:** Keep PRs atomic (one logical change). Clean up commits and rebase on `dev` to ensure no unrelated commits (e.g. from `main`) are included. Push updates to the existing PR branch instead of closing and reopening.
- [x] **Title Prefix:** To clearly categorize this pull request, prefix the pull request title using one of the following:
  - **BREAKING CHANGE**: Significant changes that may affect compatibility
  - **build**: Changes that affect the build system or external dependencies
  - **ci**: Changes to our continuous integration processes or workflows
  - **chore**: Refactor, cleanup, or other non-functional code changes
  - **docs**: Documentation update or addition
  - **feat**: Introduces a new feature or enhancement to the codebase
  - **fix**: Bug fix or error correction
  - **i18n**: Internationalization or localization changes
  - **perf**: Performance improvement
  - **refactor**: Code restructuring for better maintainability, readability, or scalability
  - **style**: Changes that do not affect the meaning of the code (white space, formatting, missing semi-colons, etc.)
  - **test**: Adding missing tests or correcting existing tests
  - **WIP**: Work in progress, a temporary label for incomplete or ongoing work

# Changelog Entry

### Description

- When a tool returns an HTML embed (via `HTMLResponse` or the tuple [(html, headers)] replaces the tool's actual return value with a hardcoded message: 
`"{tool_function_name}: Embedded UI result is active and visible to the user."`

This prevents tool authors from providing the LLM with specific context about the generated embed (e.g., parameters used, status, what was actually generated). This PR introduces support for an optional tuple (HTMLReturn, str | dict | list) that allows authors to override this generic message with the tuples second member, providing the LLM with actionable context while still rendering the embedded UI to the user.

See [discussion](https://github.com/open-webui/open-webui/discussions/22688): 

### Added

- `Support for parsing a Tuple of (HTMLResult, str | dict | list) returns of Tools within process_tool_result to pass custom tool message response to the LLM alongside the HTML embed to the UI.

### Changed

- Updated process_tool_result (open-webui/backend/open_webui/utils/middleware.py) to fall back to the generic "Embedded UI result is active..." message only if the HTMLResult is returned alone.

### Deprecated

- None

### Removed

- None

### Fixed

- None

### Security

- None

### Breaking Changes

- None

---

### Additional Information

- This change is fully backwards-compatible. Existing tools that send HTMLResponse alone will continue to see the current default behavior.

### Screenshots or Videos

[Test on wheather tool](https://github.com/user-attachments/assets/9a84b14b-10e0-4b4e-8083-710da731729d.gif)

Test on RPG tools

The LLM can call the tool show the embed and get a helpfull message on tool output for itself for future refereces in the chat

documentation updated in PR :
https://github.com/open-webui/docs/pull/1140
for the changes

### Contributor License Agreement

- [ x] By submitting this pull request, I confirm that I have read and fully agree to the [Contributor License Agreement (CLA)](https://github.com/open-webui/open-webui/blob/main/CONTRIBUTOR_LICENSE_AGREEMENT), and I am providing my contributions under its terms.

> [!NOTE]
> Deleting the CLA section will lead to immediate closure of your PR and it will not be merged in.

### CODE REVIEW

**Comment by Haervwe** [comment]
The only concern i have on this approach is that headers wont accept newlines nor very long messages, but is to my use cases at least  a good tradeoff against going a more complex route. mostly when passed on the HTMLResponse object itself and not that much of an issue if at all on the tuple path though

**Comment by Haervwe** [comment]
Ok this approach is limited by the headers limits on tools , since the tuple path dont apply to native tools, another option is to allow the return of tuple or a pydantic object that holds the HTML embed and and the user message:

as:
return HTMLResponse, "This is the actual data"

of 

class ToolResult(BaseModel:
     html_embed: HTMLResponse
     result_text: str 
     
 living on utils? ? or models. let met know if this is prefered 
 
 and returning that on tools and validating.

**Comment by Haervwe** [comment]
Ok, went all in on the Tuple approach since it seems simpler. i have tested it on different tools and its working as inteded.

IMGS:
image
image

The model creates the embed and also gets contextual information in this case the resource links and a different message sent as this :

image


i tested it with custom messages being strings, lists and objects since those are the more common cases to my knoledge.

the pattern to me seems clean.

**Comment by Haervwe** [comment]
tested regression for built in tools. terminal tools, and MCP server tools and theres no collition nor interferece.

File: backend/open_webui/utils/middleware.py [hunk 1]
```diff
@@ -966,6 +966,12 @@ def process_tool_result(
     tool_result_embeds = []
     EXTERNAL_TOOL_TYPES = ("external", "action", "terminal")
 
+    custom_message = None
+    if isinstance(tool_result, tuple) and len(tool_result) == 2 and isinstance(tool_result[0], HTMLResponse):
```
Comment (Haervwe) [inline backend/open_webui/utils/middleware.py:941]: This check safely differenciates the return type since HTMLResponse object can only come with the correct type for the assertion if returned in process by the python backend. so this check should not affect any other execution path. tested on 70+ chats with different tool combinations

**Comment by tjbck** [comment]
Addressed in dev!
